src/ess/reduce/ui.py

```diff
@@ -18,145 +20,242 @@
+) -> None:
+    def refresh_output(_: widgets.Button):
+        output_widget.clear_output()
+
+    refresh_button.on_click(refresh_output)
+
+
+def connect_output_selection_and_parameter_run_button(
+    *output_selection_widgets: widgets.Widget,
+    parameter_refresh_button: widgets.Button,
+    run_button: widgets.Button,
+) -> None:
+    # Disable run button when output selection changes
+    def observe_selection_change(_) -> None:
+        run_button.disabled = True
+        run_button.tooltip = 'To run the workflow, refresh parameters.'
+
+    for output_selection_widget in output_selection_widgets:
+        output_selection_widget.observe(observe_selection_change)
+
+    # Enable run button when parameters are generated
+    original_run_button_tooltip = run_button.tooltip
+
+    def observe_parameter_refrheshed(_) -> None:
+        run_button.disabled = False
+        run_button.tooltip = original_run_button_tooltip
+
+    parameter_refresh_button.on_click(observe_parameter_refrheshed)
+
+
+class WorkflowWidget(widgets.TwoByTwoLayout):
+    def __init__(
+        self, workflow: sl.Pipeline, result_registry: dict | None = None, **kwargs
+    ):
+        self.output_selection_box = OutputSelectionWidget(workflow)
+        registry_getter = partial(
+            _registry_getter_template, workflow, self.output_selection_box
+        )
+        self.parameter_box = ParameterBox(registry_getter)
+        workflow_runner = partial(
+            _workflow_runner_template,
+            workflow.copy,
+            self.output_selection_box,
+            self.parameter_box,
+        )
+        self.result_box = ResultBox(workflow_runner, result_registry)
+        connect_refresh_button(
+            self.parameter_box.parameter_refresh_button, self.result_box.output
+        )
+        connect_output_selection_and_parameter_run_button(
+            self.output_selection_box.typical_outputs_widget,
+            self.output_selection_box.possible_outputs_widget,
+            parameter_refresh_button=self.parameter_box.parameter_refresh_button,
+            run_button=self.result_box.run_button,
+        )
+        for box in (self.output_selection_box, self.parameter_box, self.result_box):
+            box.layout.border = '1px solid black'
+
+        super().__init__(
+            top_left=self.output_selection_box,
+            top_right=self.parameter_box,
+            bottom_left=self.result_box,
+            grid_gap="10px",
+            layout=default_layout,
+            **kwargs,
+        )
 
-def run_workflow(_: widgets.Button) -> None:
-    workflow_constructor = cast(Callable[[], sciline.Pipeline], workflow_select.value)
-    selected_workflow = workflow_constructor()
-    outputs = possible_outputs_widget.value + typical_outputs_widget.value
-    registry = get_parameters(selected_workflow, outputs)
 
-    values = collect_values(parameter_box, registry.keys())
+def workflow_widget_from_constructor(
```

Comment:
Where is this function used? Do we need it?

Reply:
It's for creating a workflow widget from the ``workflow-function`` without selecting it from the higher-widget.

It'll be easier to make a widget quickly from a notebook.

Reply:
Hm, I don't really understand. It doesn't seem easier to use `workflow_widget_from_constructor(workflow_function, result_registry)` than  `WorkflowWidget(workflow_function(), result_registry)`. But maybe I'm missing the point.

Reply:
Hmm that's true. I think it's also a legacy from the previous cleanup before I made a wrapper classes.
I thought we needed more explicit function since we already have `workflow_widget`...
I'll just delete it for now.